File: backend/src/main/java/isep/webapp/terasync/service/user/UserService.java
package isep.webapp.terasync.service.user;

import isep.webapp.terasync.model.Entity;
import isep.webapp.terasync.model.query.select.ByToken;
import isep.webapp.terasync.model.query.update.PasswordByEmail;
import isep.webapp.terasync.model.query.update.PfpImgPathByEmail;
import isep.webapp.terasync.model.query.update.TokenByEmail;
import isep.webapp.terasync.model.query.update.TokenByOldToken;
import isep.webapp.terasync.service.EntityService;
import org.springframework.data.jpa.repository.JpaRepository;

public abstract class UserService<T extends Entity, R extends JpaRepository<T, Integer>> extends EntityService<T, R> {
    public UserService(R entityRepository) {
        super(entityRepository);
    }
    public abstract T selectByEmail(String email);
    public abstract T selectByToken(ByToken byToken);
    public abstract Integer verifyEmail(String email);
    public abstract Integer updatePasswordByEmail(PasswordByEmail passwordByEmail);
    public abstract Integer updateTokenByEmail(TokenByEmail tokenByEmail);
    public abstract Integer updateTokenByOldToken(TokenByOldToken tokenByOldToken);
    public abstract Integer updatePfpImgPathByEmail(PfpImgPathByEmail pfpImgPathByEmail);
}

File: front/src/app/service/user/user.service.ts
import {HttpClient} from "@angular/common/http";
import {EntityService} from "../entity.service";
import {Injectable} from "@angular/core";
import {Observable} from "rxjs";
import {TokenByEmail} from "../../model/query/update/token-by-email";
import {TokenByOldToken} from "../../model/query/update/token-by-old-token";
import {PasswordByEmail} from "../../model/query/update/password-by-email";
import {ByToken} from "../../model/query/select/by-token";
import {User} from "../../model/user/user";
import {PfpImgPathByEmail} from "../../model/query/update/pfp-img-path-by-email";

@Injectable({
  providedIn: 'root'
})
export abstract class UserService<T extends User> extends EntityService<T> {

  protected constructor(http: HttpClient, entity: string) {
    super(http, entity);
  }

  public findUserByEmail(email: String): Observable<T> {
    return this.http.get<T>(`${this.apiBackendUrl}/${this.entityName}/select-user-by-email/${email}`);
  }

  public findUserByToken(byToken: ByToken): Observable<T> {
    return this.http.post<T>(`${this.apiBackendUrl}/${this.entityName}/select-user-by-token`, byToken);
  }

  public verifyEmail(email: String): Observable<number> {
    return this.http.get<number>(`${this.apiBackendUrl}/${this.entityName}/verify-email/${email}`);
  }

  public updatePasswordByEmail(passwordByEmail: PasswordByEmail): Observable<number> {
    return this.http.post<number>(`${this.apiBackendUrl}/${this.entityName}/update-password-by-email`, passwordByEmail);
  }

  public updateTokenByEmail(tokenByEmail: TokenByEmail): Observable<number> {
    return this.http.post<number>(`${this.apiBackendUrl}/${this.entityName}/update-token-by-email`, tokenByEmail);
  }

  public updateTokenByOldToken(tokenByOldToken: TokenByOldToken): Observable<number> {
    return this.http.post<number>(`${this.apiBackendUrl}/${this.entityName}/update-token-by-old-token`, tokenByOldToken);
  }

  public updatePfpImgPathByEmail(pfpImgPathByEmail: PfpImgPathByEmail): Observable<number> {
    return this.http.post<number>(`${this.apiBackendUrl}/${this.entityName}/update-pfp-img-path-by-email`, pfpImgPathByEmail);
  }

  public deleteUserAndPfpImg(user: T): Promise<boolean> {
    return new Promise<boolean>((resolve, reject) => {
      this.deleteEntity(user.userId!).subscribe({
        next: () => {
          console.log(`User deleted: ${user.userId!}`);
          if(user.pfpImgPath != undefined && user.pfpImgPath.length > 0) {
            this.deleteFile(user.pfpImgPath).subscribe({
              next: (response: boolean) => {
                console.log(`Pfp img deleted: ${response} for user: ${user.userId!}` );
              },
              error: (error: any) => {
                console.error(error);
              }
            });
          }
          resolve(true);
        },
        error: (error: any) => {
          console.log(`User with id ${user.userId!} can't be deleted`)
          resolve(false);
        }
      })
    });
  }
}
